From c5d9fbbfdc88974e635dd4674e710618db99c08d Mon Sep 17 00:00:00 2001
From: Colin Huang <colin.huang2@amd.com>
Date: Thu, 13 Nov 2025 01:56:03 +0800
Subject: [PATCH] fix reset button not working

meson: fix reset-button-do-warm-reboot not work
button_handler: fix hostNumStr is empty when reset button pressed.

tested: A reset button press works to reset the system.
A power button press work to power system without process restart.

Change-Id: Ifea143860e4173d10c2386fa13509a0f935a1b21
Signed-off-by: Colin Huang <colin.huang2@amd.com>
---
 meson.build            |  2 +-
 meson_config.hpp.in    |  2 ++
 src/button_handler.cpp | 48 +++++++++++++++++++++++++++++++++---------
 3 files changed, 41 insertions(+), 11 deletions(-)

diff --git a/meson.build b/meson.build
index fe8ec69..10f83fa 100644
--- a/meson.build
+++ b/meson.build
@@ -15,7 +15,7 @@ conf_data.set_quoted('ID_LED_GROUP', get_option('id-led-group'))
 conf_data.set_quoted('POWER_BUTTON_PROFILE', get_option('power-button-profile'))
 conf_data.set('LONG_PRESS_TIME_MS', get_option('long-press-time-ms'))
 conf_data.set('LOOKUP_GPIO_BASE', get_option('lookup-gpio-base').allowed())
-conf_data.set(
+conf_data.set10(
     'ENABLE_RESET_BUTTON_DO_WARM_REBOOT',
     get_option('reset-button-do-warm-reboot').allowed(),
 )
diff --git a/meson_config.hpp.in b/meson_config.hpp.in
index ffc82a6..f3c3aba 100644
--- a/meson_config.hpp.in
+++ b/meson_config.hpp.in
@@ -27,6 +27,8 @@ constexpr inline auto GPIO_BASE_LABEL_NAME = "1e780000.gpio";
 constexpr inline auto gpioDefFile = "/etc/default/obmc/gpio/gpio_defs.json";
 #define LOOKUP_GPIO_BASE @LOOKUP_GPIO_BASE@
 
+#define ENABLE_RESET_BUTTON_DO_WARM_REBOOT @ENABLE_RESET_BUTTON_DO_WARM_REBOOT@
+
 constexpr inline auto POWER_BUTTON_PROFILE = @POWER_BUTTON_PROFILE@;
 constexpr inline auto ID_LED_GROUP = @ID_LED_GROUP@;
 constexpr inline const auto LONG_PRESS_TIME_MS =
diff --git a/src/button_handler.cpp b/src/button_handler.cpp
index f18f1b5..17e4dfc 100644
--- a/src/button_handler.cpp
+++ b/src/button_handler.cpp
@@ -287,8 +287,32 @@ void Handler::handlePowerEvent(PowerEvent powerEventType,
     std::variant<Host::Transition, Chassis::Transition> transition;
 
     size_t hostNumber = 0;
-    std::string hostNumStr =
-        objectPath.substr(std::string(POWER_DBUS_OBJECT_NAME).length());
+    std::string hostNumStr = [&]() -> std::string {
+        switch (powerEventType)
+        {
+            case PowerEvent::powerReleased:
+            {
+                return objectPath.substr(
+                    std::string(POWER_DBUS_OBJECT_NAME).length() - 1);
+            }
+
+            case PowerEvent::resetReleased:
+            {
+                return objectPath.substr(
+                    std::string(RESET_DBUS_OBJECT_NAME).length() - 1);
+            }
+
+            default:
+            {
+                lg2::error("{EVENT} is invalid power event. skipping...",
+                           "EVENT", powerEventType);
+                throw std::runtime_error("Invalid power event type");
+            }
+        }
+
+        throw std::runtime_error("Failed to deduce hostNumStr from objectPath");
+    }();
+
     auto isMultiHostSystem = isMultiHost();
 
     if (hostSelectButtonMode)
@@ -314,17 +338,21 @@ void Handler::handlePowerEvent(PowerEvent powerEventType,
     {
         case PowerEvent::powerReleased:
         {
-            for (const auto& iter : multiPwrBtnActConf[stoi(hostNumStr) - 1])
+            if (!multiPwrBtnActConf.empty())
             {
-                if (duration > std::chrono::milliseconds(iter.first))
+                for (const auto& iter :
+                     multiPwrBtnActConf[stoi(hostNumStr) - 1])
                 {
-                    dbusIfaceName = chassisIface;
-                    transitionName = "RequestedPowerTransition";
-                    objPathName = CHASSIS_STATE_OBJECT_NAME + hostNumStr;
-                    transition = iter.second;
+                    if (duration > std::chrono::milliseconds(iter.first))
+                    {
+                        dbusIfaceName = chassisIface;
+                        transitionName = "RequestedPowerTransition";
+                        objPathName = CHASSIS_STATE_OBJECT_NAME + hostNumStr;
+                        transition = iter.second;
+                        break;
+                    }
                 }
             }
-            break;
 
             if (duration <= LONG_PRESS_TIME_MS)
             {
@@ -386,7 +414,7 @@ void Handler::handlePowerEvent(PowerEvent powerEventType,
             }
 
             lg2::info("Handling reset button press");
-#ifdef ENABLE_RESET_BUTTON_DO_WARM_REBOOT
+#if ENABLE_RESET_BUTTON_DO_WARM_REBOOT
             transition = Host::Transition::ForceWarmReboot;
 #else
             transition = Host::Transition::Reboot;
-- 
2.47.3

